oeps/oep-0036-microservice-routing-layer.rst

```diff
@@ -0,0 +1,89 @@
+OEP-36: Microservice Routing Layer
+===================================
+
+.. list-table::
+
+  * - OEP
+    - :doc:`oep-0036-microservice-routing-layer`
+  * - Title
+    - Microservice Routing Layer
+  * - Last-Modified
+    - 2019-04-25
+  * - Author
+    - Fred Smith <derf@edx.org>
+  * - Arbiter
+    - David Joy <djoy@edx.org>
+  * - Status
+    - Draft
+  * - Type
+    - Architecture
+  * - Created
+    - 2019-04-25
+
+Abstract
+========
+
+Proposes a key openedx component that will connect users to applications using a layer 7 (application layer/URL Based) routing framework. 
+
+Motivation
+==========
+
+Microservices are currently implemented using layer 4 (DNS) based routing. Each microservice occupies a unique hostname, with dedicated ELB or CDN infrastructure to support routing user traffic to the service. This pattern is simple to monitor, deploy, and manage, but slow and expensive to build, and regularly causes problems with cross service user information sharing. In Addition, modifying or splitting a monolithic service into microservices results in confusing URL changes and costly migrations.
+
+Specification
+=============
+
+A layer 7 routing application will be chosen that meets the following requirements:
+
+* High performance connection management capabilities, including support for slow clients or backends, and wire-speed throughput.
+* SSL termination on the router.
+* Flexible configuration backends including container service discovery, flat files, and filesystem routing.
+* Path based routing, including path subsets and regular expression matching
+
+This layer 7 routing application will be deployed on a common hostname.
+
+Rationale
+=========
+
+A layer 7 routing layer would allow us to separate the user facing URL or API endpoint from the backend implementation. It would allow microservices to exist that only manage a small number of endpoints, without requring developers to build out load balancing or CDN infrastructure. 
+
+We are currently using a layer 7 routing framework for the edx.org marketing site, allowing them to replace drupal generated pages with gatsby generated pages quickly and seamlessly.
+
+This change will allow developers, openedx operators and resellers to deploy new openedx sites quicker and easier than they currently can.  The router will provide a map for people to understand which services they need to deploy.
+
+
+
+Implementation
+==============
+
+* to be determined
+
+Future
+======
+
+Possible additional functionality that may be implemented at the l7 router:
+
+* Authentication
+* Rate limiting
+* Web Application Firewall
+
+
+Backwards Compatibility
+=======================
+
+Services will begin to move API and user endpoints behind this layer 7 router after declaring their top level routing path. eg.
+
+* the LMS would occupy /courses
+* prospectus would occupy /course, /school, /micromasters
+* ecommerce would occupy /orders, /payments
+* studio would occupy /studio 
+
```

Comment:
If we're moving around a bunch of URLs with 301 redirects, is it too much to force each service to only have its own top level namespace? Allowing each IDA to have multiple, arbitrary top level URLs like /course vs. /courses sounds like it will get confusing over time (as we get dozens of these), and potentially lead to name collisions and potentially configuration issues down the road. Whereas if there's a 1:1 mapping of IDA to top level, then once an IDA is established, they control everything below that namespace and they don't have to touch the centralized routing layer config to add a new directory somewhere.

Reply:
I guess the larger version of this is that I would like to better understand what the restrictions on usage are, and how we're going to reduce the chances of teams accidentally stepping on each other. I strongly support having some routing layer as proposed in this document, but I'm concerned that it will gradually grow in complexity into a hodgepodge of global and local rules (e.g. "hey, did my new redirect pattern affect images for XYZ other service?") that people are afraid to touch. I'm concerned that we're going to contort it to accommodate use cases that add complexity rather than forcing URL restrictions or requiring downstream changes. I'm concerned that someone else is going to want `/courses/`, and so we'll write a regex matcher that says `/courses/{uuid-pattern}` goes to Publisher while `/courses/{course-key}` goes to the LMS.

Reply:
Okay. Just to check then -- you think it's reasonable to have a 1:1 mapping of URL prefix to service if we're not using technical words (e.g. `/learning` instead of `/lms`)? Or do you think that we need the flexibility to give services multiple top-level directories and decouple URLs from their backing services entirely?

Reply:
> you think it's reasonable to have a 1:1 mapping of URL prefix to service if we're not using technical words (e.g. /learning instead of /lms)?

Yes.